numpy/core/src/npymath/npy_math_internal.h.src

```diff
@@ -845,3 +851,30 @@ npy_rshift@u@@c@(npy_@u@@type@ a, npy_@u@@type@ b)
 }
 /**end repeat1**/
 /**end repeat**/
+
+
+/**begin repeat
+ *
+ * #type  = ubyte, ushort, uint, ulong, ulonglong#
+ * #STYPE = BYTE,  SHORT,  INT,  LONG,  LONGLONG#
+ * #c     = ,      ,       ,     ll,    ll#
+ */
+#undef MAGIC_ARRAY
+#if 0
+/**begin repeat1
+ * #len = 8, 16, 32, 64#
+ */
+#elif NPY_BITSOF_@STYPE@ == @len@
+    #define MAGIC_ARRAY B@len@
+/**end repeat1**/
+#endif
+
+NPY_INPLACE npy_@type@
+npy_@type@_countbits(npy_@type@ a)
+{
+    a = a - ((a >> 1) & MAGIC_ARRAY[0]);
+    a = ((a & MAGIC_ARRAY[1])) + ((a >> 2) & MAGIC_ARRAY[1]);
+    return (((a + (a >> 4)) & MAGIC_ARRAY[2]) * MAGIC_ARRAY[3]) >> (NPY_BITSOF_@STYPE@ - 1)*8;
+}
+
```

Comment:
What happened to using `__builtin_popcount`? Here's how CPython does it:

https://github.com/python/cpython/blob/7569c0fe91dfcf562dee8c29798ecda74d738aa8/Include/internal/pycore_bitutils.h#L86-L130

Reply:
Yeah, thanks for bringing it up. So there is GNU GCC and Clangs in-built popcount. Also there is MSVC. Is there any other popcount that is popular enough that we have to include?

Reply:
I think we can take the SSE and AVX in a follow-up commit, they to have efficient popcounts

Reply:
I have taken a note to add MSVC support

Reply:
Added in 0ac9918(clang/gcc) and 8a9b6ef(Windows)